homeassistant/helpers/storage.py

```diff
@@ -0,0 +1,159 @@
+        self.version = version
+        self.key = key
+        self.hass = hass
+        self._data = None
+        self._unsub_delay_listener = None
+        self._unsub_stop_listener = None
+        self._write_lock = asyncio.Lock()
+
+    @property
+    def path(self):
+        """Return the config path."""
+        return self.hass.config.path(STORAGE_DIR, self.key)
+
+    async def async_load(self):
+        """Load data.
+
+        If the expected version does not match the given version, the migrate
+        function will be invoked with await migrate_func(version, config).
+        """
+        if self._data is not None:
+            data = self._data
+        else:
+            data = await self.hass.async_add_executor_job(
+                json.load_json, self.path, None)
+
+            if data is None:
+                return {}
+
+        if data['version'] == self.version:
+            return data['data']
+
+        return await self._async_migrate_func(data['version'], data['data'])
+
+    async def async_save(self, data: Dict, *, delay: Optional[int] = None):
+        """Save data with an optional delay."""
+        self._data = {
+            'version': self.version,
+            'key': self.key,
+            'data': data,
+        }
+
+        self._async_cleanup_delay_listener()
+
+        if delay is None:
+            self._async_cleanup_stop_listener()
+            await self._handle_write_data()
+            return
+
+        self._unsub_delay_listener = async_call_later(
+            self.hass, delay, self._async_callback_delayed_write)
+
+        self._async_ensure_stop_listener()
+
+    @callback
+    def _async_ensure_stop_listener(self):
+        """Ensure that we write if we quit before delay has passed."""
+        if self._unsub_stop_listener is None:
+            self._unsub_stop_listener = self.hass.bus.async_listen_once(
+                EVENT_HOMEASSISTANT_STOP, self._async_callback_stop_write)
+
+    @callback
+    def _async_cleanup_stop_listener(self):
+        """Clean up a stop listener."""
+        if self._unsub_stop_listener is not None:
+            self._unsub_stop_listener()
+            self._unsub_stop_listener = None
+
+    @callback
+    def _async_cleanup_delay_listener(self):
+        """Clean up a delay listener."""
+        if self._unsub_delay_listener is not None:
+            self._unsub_delay_listener()
+            self._unsub_delay_listener = None
+
+    async def _async_callback_delayed_write(self, _now):
+        """Handle a delayed write callback."""
+        self._unsub_delay_listener = None
+        self._async_cleanup_stop_listener()
+        await self._handle_write_data()
+
+    async def _async_callback_stop_write(self, _event):
+        """Handle a write because Home Assistant is stopping."""
+        self._unsub_stop_listener = None
+        self._async_cleanup_delay_listener()
+        await self._handle_write_data()
+
+    async def _handle_write_data(self, *_args):
```

Comment:
you use for all async function `async`. That is the only function they use a other name concept